### PR TITLE
Added: support for an extra basis in the mixed formulations

### PR DESCRIPTION
**Pull request by akva2**
This contains preparation work and then support for the extra basis in the mixed formulations.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=270 ifem-navierstokes=220 ifem-boussinesq=117 please

File: src/ASM/ASMs1D.C [hunk 1]
```diff
@@ -1956,15 +1956,14 @@ bool ASMs1D::getNoStructElms (int& n1, int& n2, int& n3) const
 bool ASMs1D::evaluate (const FunctionBase* func, RealArray& values,
                        int, double time) const
 {
-  Go::SplineCurve* scrv = SplineUtils::project(curv,*func,func->dim(),time);
+  std::unique_ptr<Go::SplineCurve> scrv(SplineUtils::project(curv.get(),*func,func->dim(),time));
```
Comment (kmokstad) [inline src/ASM/ASMs1D.C:1959]: Don't see what is gained here by the template usage (except for the more lengthy statement), since this pointer is just a local variable which is deleted before returning.

File: src/ASM/ASMs1D.h [hunk 2]
```diff
@@ -466,8 +466,8 @@ class ASMs1D : public ASMstruct, public ASMunstruct, public ASM1D
   virtual void getElmConnectivities(IntMat& neighs) const;
 
 protected:
-  Go::SplineCurve* curv; //!< Pointer to the actual spline curve object
-  Go::SplineCurve* proj; //!< Pointer to spline curve for projection basis
+  std::shared_ptr<Go::SplineCurve> curv; //!< Pointer to the actual spline curve object
+  std::shared_ptr<Go::SplineCurve> proj; //!< Pointer to spline curve for projection basis
```
Comment (kmokstad) [inline src/ASM/ASMs1D.h:470]: Then remove "Pointer to" in the doxy to keep it within 80 columns.

Applies elsewhere in similar declarations as well.

File: src/ASM/ASMs2D.C [hunk 3]
```diff
@@ -155,16 +154,14 @@ bool ASMs2D::read (std::istream& is)
...
+    surf.reset();
     return false;
   }
   else if (surf->dimension() < 2)
   {
     std::cerr <<" *** ASMs2D::read: Invalid spline surface patch, dim="
-	      << surf->dimension() << std::endl;
-    delete surf;
-    surf = nullptr;
+          << surf->dimension() << std::endl;
```
Comment (kmokstad) [inline src/ASM/ASMs2D.C:163]: Wrong indent

File: src/ASM/LR/ASMu3Dmx.h [hunk 4]
```diff
@@ -36,7 +36,7 @@ class ASMu3Dmx : public ASMu3D, private ASMmxBase
   //! \brief Copy constructor.
   ASMu3Dmx(const ASMu3Dmx& patch, const CharVec& n_f = CharVec(3,0));
   //! \brief Empty destructor.
-  virtual ~ASMu3Dmx() {}
+  virtual ~ASMu3Dmx() = default;
```
Comment (kmokstad) [inline src/ASM/LR/ASMu3Dmx.h:39]: Guess these are equivalent, right? Just to be consistent.

File: src/ASM/ASMs2Dmx.C [hunk 5]
```diff
@@ -1045,23 +1045,27 @@ bool ASMs2Dmx::evalSolutionPiola (Matrix& sField, const Vector& locSol,
...
 
-  size_t len = std::accumulate(nb.begin(),nb.end(),0u);
+  size_t len = std::inner_product(nb.begin(), nb.end(), nfx.begin(), 0u);
   bool withPressure = len == locSol.size();
+  size_t nFields;
   if (!withPressure)
-    len = nb[0] + nb[1];
+    len = nb[0] + nb[1], nFields = 2;
+  else
+    nFields = std::accumulate(nfx.begin(), nfx.end(), 0);
```
Comment (kmokstad) [inline src/ASM/ASMs2Dmx.C:1074]: Maybe better with
```
  size_t nFields = 2;
  if (withPressure)
    nFields = std::accumulate(nfx.begin(), nfx.end(), 0u);
  else
    len = nb[0] + nb[1];
```
The logic is in any case hard to grasp

File: src/ASM/LR/ASMu2Dmx.C [hunk 6]
```diff
@@ -900,37 +900,41 @@ bool ASMu2Dmx::evalSolutionPiola (Matrix& sField, const Vector& locSol,
...
-  size_t len = std::accumulate(nb.begin(),nb.end(),0u);
+  size_t len = std::inner_product(nb.begin(), nb.end(), nfx.begin(), 0u);
   bool withPressure = len == locSol.size();
+  size_t nFields;
   if (!withPressure)
-    len = nb[0] + nb[1];
+    len = nb[0] + nb[1], nFields = 2;
+  else
+    nFields = std::accumulate(nfx.begin(), nfx.end(), 0);
+
```
Comment (kmokstad) [inline src/ASM/LR/ASMu2Dmx.C:930]: See previous comment for ASMs2Dmx

**Review by kmokstad — APPROVED** [review]
Although I'm not too fond of "smart" pointers, I guess this is fine. Did not go into the logic though which is hard to follow, and now will be even harder since we don't need to understand where the dynamic objects go out of scope and needs to be deleted (but that' my personal view only).

File: src/ASM/LR/ASMu3Dmx.h (hunk 4 above)
Comment (akva2) [reply src/ASM/LR/ASMu3Dmx.h:39]: yes, they are equivalent.

File: src/ASM/ASMs2D.C (hunk 3 above)
Comment (akva2) [reply src/ASM/ASMs2D.C:163]: fixed. my IDE replaced a tab with spaces but my tabs are 2, not 8..

File: src/ASM/ASMs1D.h (hunk 2 above)
Comment (akva2) [reply src/ASM/ASMs1D.h]: ok.

File: src/ASM/ASMs2Dmx.C (hunk 5 above)
Comment (akva2) [reply src/ASM/ASMs2Dmx.C:1072]: okay, it will cause sca warnings but those are harmless of course.

**Comment by akva2** [comment]
jenkins build this with downstreams please

File: src/ASM/ASMs1D.C (hunk 1 above)
Comment (akva2) [reply src/ASM/ASMs1D.C:1959]: i now immediately see that this pointer won't leak, as opposed to having to look through the code to find the matching delete, and making sure there are not early exits or the likes where it would leak. in this case, sure, there's none of those cases but in general i think that new and delete should be removed from the user facing language and be reserved to low level libraries.

**Comment by akva2** [comment]
jenkins build this with downstreams please
